facebook/react-native
# RCTWebView bridge redirection error fix

Fixing RCTWebView redirection error 


## Motivation
Inspired by following tickets: [10941](https://github.com/facebook/react-native/pull/10941), [10865](https://github.com/facebook/react-native/issues/10865)
On iOS, when working with RCTWebView and bridging between react native and native JavaScript (i.e: _isMessagingEnabled == YES)
any redirection done on the JavaScript side raises the following error on the dev env: 
`Setting onMessage on a WebView overrides existing values of window.postMessage, but a previous value was defined.`

![react-native-bug](https://user-images.githubusercontent.com/7502854/31894156-3ca7af96-b816-11e7-9946-593549614292.png)

Like stated above, this error occurs when the `onMessage` prop in the react native side is set and used.
In order to understand this error and why it happens, one must understand two methods
That The native iOS `UIWebView's` delegate has
#### 1) - (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType
This method is called each time a redirection occurs on the browsers' JS side, and not just for the main window (i.e: for iframes as well; keep that part in mind).
If the method returns `YES` and only then will the WebView continue to execute the redirection load
#### 2) - (void)webViewDidFinishLoad:(UIWebView *)webView
Each successful load (be it a redirection or a page load) will call this method, pretty simple.

### The Problem
When having a bridge between RN and native JS, the `webViewDidFinishLoad` method
performs an extra check to validate if the `window.postMessage` was left untouched and raises the above mentioned error if it was.
The method then overrides the `window.postMessage` after saving it under`window.originalPostMessage`
i.e:
`window.originalPostMessage = window.postMessage;
window.postMessage = function(data) { /* New code */}
`
While this solution is sufficient for cases where there is a simple one page bridge, it doesn't cover cases where these delegate methods will be called again due to any kind of redirection that doesn't kill the last modified window (i.e: iframes).
This solution only takes care of cases where the main window changes (navigating to a new page).
If these delegate methods will be called again, the first dev check will fail, showing the error; and the `window.originalPostMessage` will be overriding again, deleting the original `window.postMessage` implementation

The applied fix adds two simple checks:
1. Take `window.originalPostMessage` method into consideration and check whether it has the implementation of the `window.postMessage` method (dev env only)
2. Check if the `window.originalPostMessage` exists before reassigning it with a new value


## Test  Plan
1. populate the WebView onMessage prop in react native
2. load the WebView from a custom html
3. add an iframe to the html

behavior before fix: red error screen will be shown
expected behavior/after fix: everything will continue working normally as it should

## Release Notes
